src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1037,12 +1037,17 @@ public static double Min(double val1, double val2)
             // This matches the IEEE 754:2019 `minimum` function
             //
             // It propagates NaN inputs back to the caller and
-            // otherwise returns the larger of the inputs. It
-            // treats +0 as larger than -0 as per the specification.
+            // otherwise returns the lesser of the inputs. It
+            // treats +0 as lesser than -0 as per the specification.
 
-            if (val1 != val2 && !double.IsNaN(val1))
+            if (val1 != val2)
             {
-                return val1 < val2 ? val1 : val2;
+                if (!double.IsNaN(val1))
```

Comment:
Just curious, when we write C# that's potentially in a hot path is there any value in putting "most likely branch first" (ie., "test is most likely true")? This would mean that the codegen will maintain this ordering + the CPU in the absence of other information will assume the branch is true + this isn't overwhelmed by accurate subsequent branch prediction?

I only ask because I'd personally find something like this easier to grok if the double.IsNaN was not negated.

Reply:
That's why the code is how it is. With the way the C# emits IL and the JIT compiles it, then by default (that is when no PGO data is present), the code is effectively:
```csharp
if (cond)
{
    // predicted
}

// unpredicted
```

It's the opposite for ternaries:
```
cond ? unpredicted : predicated
```

This ordering ensures that the most predicted path is the `return val1 < val2 ? val1 : val2;` scenario which makes a decently measurable impact on perf. We could potentially improve our PGO data here, but since the `Math` methods are generally always considered "hot", its worth micro-optimizing anyways.

Reply:
I think the BCL is too large to have accurate PGO data for many of these functions, there are likewise cases where PGO data will mess up or hinder normal workflows.

Many of the `Math` functions in particular are things which should be intrinsic or considered "always hot". They are also functions where the default predicted case and the actual case often don't line up. Many native compilers explicitly treat them as `50/50` and "always inline" in which case the PGO data is irrelevant.

Reply:
We change the code layout and branch optimization logic in the jit pretty frequently and more changes are in the plans for the future. I don't know how to reconcile that with what you are saying.

Reply:
We make tweaks yes, but even then the JIT typically defaults to (barring PGO data telling it otherwise), matching the order so that `if (cond) { predicted } unpredicted`. Part of this is just because its the implied execution order but also because it produces the smallest (and generally fastest) code.

Changing this causes the JIT to do more work because it either has to insert more branches or rearrange blocks and that's not worth doing unless we have data (like PGO) indicating otherwise.

Functions like `Math.Min` are core functions that are decently hot and so having them ordered such that the default codegen is better, is general goodness. Changing it would show up in our weekly perf triage and result in it getting reverted back. If we wanted to change it, we'd need additional functionality so that we could statically tell the JIT the "intended" behavior.